### DISCUSSION TITLE
Hardware Connection Help

### DISCUSSION THREAD

**Discussion by ClemensElflein** [xtech/self-o-mat#18] [General]
@maehw 

if you need infos how to get the hardware to connect, just write here.

some things I just remembered:

- the self-o-mat software looks for a config file in `/opt/.selfomat.type`. The file needs to contain a string for the type of box that is currently running. Then the software loads the box specific settings file from the ./settings folder in this repository.
- In the settings file the serial port name can be set. You probably need to use /dev/ttyUSB0.
- There is an option in the web interface to set the Neopixel LED count as well as the data format (BGR, RGB). You can enable it by setting show_led_setup to true in the settings file described above.

**Comment by maehw** [comment]
OK, great.
The first option is already taken care of in the wiki (https://github.com/xtech/self-o-mat/wiki/Configuration#settings).

Bringing up the Arduino hardware & firmware will be documented here: https://github.com/xtech/self-o-mat/wiki/Controller-board

Either I don't understand this piece of code or if not needed, the `PIN_SWITCH` pin should be pulled to GND and to 5V, right?

```
  // If the switch is on, turn on instantly BUT ONLY THIS FIRST TIME
  if(digitalRead(PIN_SWITCH) == LOW) {
    // We want to call exit function but not enter function
    currentState = &OffState::INSTANCE;
    goToState(&BootingState::INSTANCE);
  } else {
    goToState(&OffState::INSTANCE);
  }
```

**Reply by ClemensElflein** [reply]
You are correct, the PIN_SWITCH needs to be pulled to GND. Sorry!

Normally the system starts in the "Off" state, but after an update the controller will be reset and run the setup() once again. This state needs to be catched since we don't want the board to cut the power after an update. Therefore we check in the setup() if the switch is enabled (= pin pulled LOW, as you correctly noted) and we switch the system to booting state instantly (Booting state will have power enabled).

**Reply by ClemensElflein** [reply]
I have updated the wiki and my previous pinout. I am currently unable to test the hardware myself, therefore don't hesitate to ask quickly if anything is unclear.

**Comment by maehw** [comment]
Hi,

could you please add some information about command encoding? If I get it right, it's not only plain characters being sent over the line:

https://github.com/xtech/self-o-mat/wiki/Controller-board#serial-commands

Edit: BTW, I've added a state diagram for the SelfomatController's firmware.

Cheers

**Comment by ClemensElflein** [comment]
Thank you for your work!

The communication between the Arduino and the host software is basically COBS encoded [Wiki](https://en.wikipedia.org/wiki/Consistent_Overhead_Byte_Stuffing). The only difference is that we can't use the 0 byte as delimiter, because of issues with the boot loader. Therefore I used the space character as delimiter. For decoding and encoding I am using the PacketSerial arduino library [https://www.arduino.cc/reference/en/libraries/packetserial/](https://www.arduino.cc/reference/en/libraries/packetserial/).

With the encoding done, we can talk about the packet format.
Basically each sent packet starts with a command byte and optionally some data bytes. The receiver reads the first byte and then decides how many data bytes are needed as an argument (for many commands no data is needed). Then it checks if enough data is present and reads the data.

For example the "accept agreement" command simply consists of the command 'a' followed by a space character ' '.

The "write settings" command consists of a '$' byte and is followed by sizeof(settings) bytes of data. The settings datatype is a struct which holds all settings of the self-o-mat controller board, the last two bytes in the struct are used as CRC checksum. Finally a space character ' ' needs to be sent in order to finalize the packet.

Nice work on the state diagram, but it does seem to miss some transitions (we can not get back to the OffState for example). Additionally the first transition is wrong: PIN_SWITCH == HIGH should go to OffState whereas PIN_SWITCH == LOW should transition to the BootingState.

**Comment by ClemensElflein** [comment]
BTW did you get the connection to work? Can you now use the software as intended (live camera preview, trigger photo, web app)?

**Reply by maehw** [reply]
Hi Clemens,

thank you for your messages.

I've fixed the first transition in the state diagram.

I don't quite understand your comment "miss some transitions (we can not get back to the OffState for example)" - are there transitions missing, wrong or too many?

The following code snippet of `OffState.cpp` made me think there is a transition out of the `OffState`:

```
  if (digitalRead(PIN_SWITCH) == LOW) {
    // The switch is on and we're allowed to exit the off state->do it
    return &BootingState::INSTANCE;
  }
```

I am currently on the road (rather travelling in a train).

Live camera preview has been working. Trigger by mouse click (later touch screen) also works.. but as there's no software countdown (yet), I'd like to go with the Arduino in between as I've everything prepared. Taking photos by my camera also has worked and showing and storing them as well. I haven't taken any look at th web app yet. I'd like the Arduino + Raspberry Pi set up running. But I also haven't really touched the serial reader code so far.

It would be great to test both components separately - i.e. either mock the Arduino to see if the rest is working correctly or mocking the Raspberry Pi (host computer) in order to see if the Arduino works correctly.

**Comment by ClemensElflein** [comment]
> Hi Clemens,
> 
> thank you for your messages.
> 
> I've fixed the first transition in the state diagram.
> 
> I don't quite understand your comment "miss some transitions (we can not get back to the OffState for example)" - are there transitions missing, wrong or too many?

Yes, there are transitions missing from ShutDownState (and all other states which return true for needsHeartbeat()) to the OffState. It happens on heartbeat timeout. The code for this is found in the [main file in line 115](https://github.com/xtech/self-o-mat/blob/e0fe08661cbe768ee503f200e4e6763c3938daf8/Arduino/SelfomatController/SelfomatController.ino#L115).


> I am currently on the road (rather travelling in a train).
> 
> Live camera preview has been working. Trigger by mouse click (later touch screen) also works.. but as there's no software countdown (yet), I'd like to go with the Arduino in between as I've everything prepared. Taking photos by my camera also has worked and showing and storing them as well. I haven't taken any look at th web app yet. I'd like the Arduino + Raspberry Pi set up running. But I also haven't really touched the serial reader code so far.

I'm glad to hear that you are making progress! Hopefully you'll get a nice working photo booth soon!

> 
> It would be great to test both components separately - i.e. either mock the Arduino to see if the rest is working correctly or mocking the Raspberry Pi (host computer) in order to see if the Arduino works correctly.

Yes, that would be nice, although I'm not sure if it's worth the additional work.

**Comment by maehw** [comment]
Thanks for the update and for pointing me to those transitions - I really had overlooked them as I was concentrating on the state-specific code. The state diagram is updated and getting messy. It may need some improvement but it seems to be a good starting point.

I agree with the estimated additional effort it would take for the component testing approach.
